control-plane/csi-driver/src/bin/node/dev.rs

```diff
@@ -60,6 +60,9 @@ pub(crate) trait Attach: Sync + Send {
     async fn find(&self) -> Result<Option<DeviceName>, DeviceError>;
     /// Fixup parameters which cannot be set during attach, eg IO timeout
     async fn fixup(&self) -> Result<(), DeviceError>;
+    /// Get basic connections parameters like target address, transport and port,
+    /// for this attach object.
+    async fn get_conn_params(&self) -> Result<HashMap<String, String>, DeviceError>;
```

Comment:
Should this be an output of `attach`? Image if for example the attach ever tries multiple protocols, there'd be no way of knowing which one was the "winner".
Also we can use an associated type here to return exact information.

Reply:
Ideally `attach` should be doing what it is told to, after all decisions are already done w.r.t transport, so we won't be changing that part of interface. I can make this info return as part of attach though. Preferred a separate interface as it has clearly defined purpose e.g. like `parse_parameters`.

Reply:
Alright, then in that case let's keep it as a separate function then. Maybe can simplify into something like this:

```suggestion
    fn attach_parameters(&self) -> Self::Parameters;
```

Reply:
hmm, it's getting troublesome to use associated type in Attach trait because there are places e.g in `impl Device::parse()` where we return the trait as `Box<dyn Attach>` , which complains about associated type ambiguity then. To handle that will need much changes by using generics. I can rather define a concrete `struct AttachParameters` and use that. It compiles fine.

```
#[derive(Default)]
pub struct AttachParameters {
    pub _host: String,
    pub _port: u16,
    pub transport: TrType,
}

#[tonic::async_trait]
pub(crate) trait Attach: Sync + Send {
[...]
    fn attach_parameters(&self) -> AttachParameters;
}
```

Reply:
Ah that's right.
Then I'd say perhaps make it an enum:
```rust
enum AttachParameters {
  Nbd,
  Iscsi,
  Nvmf(NvmfAttachParameters),
}
```

Reply:
done now.